src/java/com/zimbra/bp/BulkDownloadServlet.java

```diff
@@ -31,7 +31,7 @@ public void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOExc
         try {
             // check the auth token
             AuthToken authToken = ZimbraServlet.getAdminAuthTokenFromCookie(req, resp);
-            if (authToken == null || !authToken.isAdmin()) {
+            if (authToken == null || !AuthToken.isAnyAdmin(authToken)) {
```

Comment:
I worry that this check may not be strict enough.  We need to be sure that admin rights are granted for the target resources and not some other resources.  For example, see how ZimbraSoapContext does more work to check what is allowed.

Reply:
I also share @fatkudu 's concern.  There also seems to be a lack of sanitization on input parameters (e.g. fileId).

I'm wondering if this extension needs a full review to see how well Delegated Admin was considered in the initial design, however I don't have the time to dig into that right now.  You may want to discuss this more with @fatkudu and @grishick as I think they've both spent some time digging into similar issues in the past.

Reply:
**Check for rights during importing account records**
If a delegated admin is allowed to access "Account Migration" page, the admin can run "Migration Wizard" without any right. (Note: createMigrationTask right is just for importing account DATA using IMAP)
On the other hand, some rigths like createAccount and createTopDomain are checked during "import account records" action on MBS.
If the admin does not have enough right(s), the action fails.

**Downloading CSV**
The result csv file is created when the import account record is succeeded, 
The file name contains AccountId and random string defined as fileToken:

    [com.zimbra.bp.BulkImportAccounts]
    String fileToken = Double.toString(Math.random() * 100);
    String outSuccessFileName = String.format("%s%s_bulk_report_%s_%s.csv", LC.zimbra_tmp_directory.value(),
        File.separator, zsc.getAuthtokenAccountId(), fileToken);

An admin need to specify the random key as fileId like this:

    https://SERVER/service/extension/com_zimbra_bulkprovision/bulkdownload?action=getBulkFile&fileID=47.46394853406404&fileFormat=reportcsv

AccountId is also checked using AuthToken:

    [com.zimbra.bp.BulkDownloadServlet]
    } else if (pFileFormat.equalsIgnoreCase(ZimbraBulkProvisionExt.FILE_FORMAT_BULK_IMPORT_REPORT)) {
        bulkFileName = String.format("%s%s_bulk_report_%s_%s.csv", LC.zimbra_tmp_directory.value(),
                File.separator, authToken.getAccountId(), pFileId);
        clientFileName = "accounts_report.csv";

The file is deleted once the file is accessed:

    try {
        is.close();
        File file = new File(bulkFileName);
        file.delete();
    } catch (Exception e) {
        ZimbraLog.webclient.error(e);
    }

Or it is deleted by crontab by default.

    40 2 * * * /opt/zimbra/libexec/zmcleantmp

The random string works as one time token. It is difficult for a third party to guess the random token and the admin's AuthToken.
Of course, the file can be accessed by a third party if HTTPS connection is decrypted.

Reply:
@k-kato I think you are suggesting that the Admin Client will somehow do the security checks by stopping domain admins seeing the Account Migration page.  If true, that simply isn't good enough to stop bad actors who can use the REST API directly.  The server API handler needs to check that authenticated user has the rights to do what they are trying to do.

Reply:
I don't think about UI side. I understand what you mean. Checking a right is important to handle API on a server.

BulkDownloadServlet.doGet handles two actions; (1)getSR (i.e. get search result) and (2)getBulkFile (i.e. get csv).
Regarding (1), rights are checked during the search action in SearchResults class. The pull request does not cause negative effect.

Regarding (2), I think there is no necessary right to download csv.
The reason is,
- any admin can call BulkImportAccountsRequest to import account records at this point,
- some rights like createAccount and createTop(Sub)Domain are checked during the import,
- an admin who calls BulkImportAccountsRequest should have a right to get the result csv file (Note: "right" is not a right defined on zimbra but a general word),
- only the admin who calls BulkImportAccountsRequest can access the result csv file as I mentioned in previous comment,
- if a bad actor uses the REST API without calling BulkImportAccountsRequest, a server just returns "File not found"

In addition, a right to run the import and/or to download a csv file is not defined on zimbra at this point.
If you think the extension must check rights to run the import and/or to download csv, we need to start from defining new right(s). It's better to address it in another ticket, I think.

Reply:
I would feel more comfortable if the same checks that BulkImportAccountsRequest does were done here too.  If I understand you correctly - this API gets the results of the BulkImportAccountsRequest - so I think it makes sense to require the same rights to get the result of the request as to make the request.

Reply:
In addition to @fatkudu's comments, can we please filter/restrict data allowed in the request param ```fileId```? As it is right now, one could put in something like ```../../../blahblah``` or perhaps nulls, etc. and we wouldn't reject it. It seems like that should be to alphanumeric and it's length needs to be limited as well. We may want to see if there are others

Reply:
@fatkudu I rethink what right is good for BulkImportAccountsRequest and for getting the result. New right(s) may need to be defined. I don't think it is suitable to check rights "createAccount", "createTopDomain" and "createSubDomain".

@plobbes I will add a logic to check the format and length of fileID.
Note:
- "action" and "fileFormat" parameter are used for switching an action. "fileID" is used for searching a target csv file. They are not used in a response from the server.
- If a wrong value is set in fileID, a server returns "File not found". Directory traversal is not caused by a wrong fileID in the current implementation.

      bulkFileName = String.format("%s%s_bulk_report_%s_%s.csv", LC.zimbra_tmp_directory.value(),
              File.separator, authToken.getAccountId(), pFileId);

      The file path with default localconfig is 
         /opt/zimbra/data/tmp/_bulk_report_[accountId]_[fileId].csv 
      Even if fileId is [xxx.csv/../../../../../yyy], the path is
         /opt/zimbra/data/tmp/_bulk_report_[accountId]_xxx.csv/../../../../yyy
      and then FileInputStream throws FileNotFoundException because "_bulk_report_[accountId]_xxx.csv" is not a directoy.